### PR TITLE
delaunay unit tests, expose is_delaunay, templates, PlainObject -> Matrix

### PR DESCRIPTION
**Pull request by alecjacobson**
This adds unit tests for the Delaunay triangulation function and exposes the `is_delaunay` function, with an overload that computes "Delaunay-ness" per (half-)edge. Along the way, I fixed some PlainObjectBase to MatrixBase and added necessary templates.

#### Check all that apply (change to `[x]`)
- [x] All changes meet [libigl style-guidelines](https://libigl.github.io/style-guidelines/).
- [x] Adds new .cpp file.
- [x] Adds corresponding unit test.
- [ ] Adds corresponding python binding.
- [ ] This is a minor change.

### CODE REVIEW

**Comment by alecjacobson** [comment]
Trying again (https://github.com/libigl/libigl/pull/955#issuecomment-431197258). I think this worked correctly.

File: tests/include/igl/copyleft/cgal/order_facets_around_edges.cpp [hunk 1]
```diff
@@ -127,23 +121,23 @@ TEST(copyleft_cgal_order_facets_around_edges, DuplicatedFaces) {
...
+//    Eigen::MatrixXi F(6, 3);
+//    F << 0, 1, 2,
+//         1, 2, 0,
+//         2, 1, 3,
+//         1, 3, 2,
+//         1, 2, 4,
+//         4, 1, 2;
+//
+//    assert_order(V, F, 1, 2, {1, 0, 2, 3, 5, 4});
+//}
```
Comment (qnzhou) [inline tests/include/igl/copyleft/cgal/order_facets_around_edges.cpp]: This seems to be undoing the bug fix from https://github.com/libigl/libigl/pull/944.  Did you pull dev?

File: include/igl/is_delaunay.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  D.setConstant(F.rows(),F.cols(),false);
+  const auto D_at = [&D,&num_faces](const int he)->typename DerivedD::Scalar&
+  {
+    const int f = he%num_faces;
+    const int c = he/num_faces;
+    return D(f,c);
+  };
+  typedef typename DerivedV::Scalar Scalar;
+  // Should use Shewchuk's predicates instead.
+  const auto float_incircle = [](
```
Comment (qnzhou) [inline include/igl/is_delaunay.cpp]: This is an inexact predicate?  Shewchuk's exact predicates is not easily templated, which does not fit the header-only convention.  What about Geogram's predicates?

Comment (jdumas) [reply include/igl/is_delaunay.cpp]: Geogram's predicate are also not header-only (but should be easier to isolate, see the [predicate psm](http://alice.loria.fr/software/geogram/doc/html/geogram_PSM.html)). But why not treat it as an external dependency that should be compiled as a pair of .cpp/.h? (Like we do for `stb_image` for example.)

File: tests/include/igl/copyleft/cgal/order_facets_around_edges.cpp (hunk 1 above)
Comment (alecjacobson) [reply tests/include/igl/copyleft/cgal/order_facets_around_edges.cpp]: Fixed.

**Comment by jdumas** [comment]
Shall we merge #961 before merging any new test-related PR? Just so that have a common framework to work with.

**Comment by jdumas** [comment]
Ah well I guess not then :D

**Pull request merged**